### ISSUE TITLE
Mistake in Finagle documentation?

### ISSUE THREAD

**Issue by scf37** [twitter/finagle#583]
http://twitter.github.io/finagle/guide/Clients.html
```Requests timed out by the Request Timeout module are not retried by default given it’s not known whether or not they were written to the wire.```

`CommonParams.withRequestTimeout` javadoc:
``` * == Client's Request Timeout ==
   *
   * The client request timeout is the maximum amount of time given to a single request
   * (if there are retries, they each get a fresh request timeout). The timeout is applied
   * only after a connection has been acquired. That is: it is applied to the interval
   * between the dispatch of the request and the receipt of the response.
```

So, will request failed by timeout is to be retried or not?

**Comment by mosesn** [comment]
@scf37 it's not retried.  The distinction is that if a request is failed for a reason that makes it eligible for automatic retrying (like it gets nacked) then it gets a fresh request timeout.

Would you like to make a PR to help clarify this?

**Comment by scf37** [comment]
@mosesn I'm not that confident in my english :-)
```
The client request timeout is the maximum amount of time given to a single request
(if there are retries, they each get a fresh request timeout). The timeout is applied
only after a connection has been acquired. That is: it is applied to the interval
between the dispatch of the request and the receipt of the response.
Requests timed out because of this parameter are not retried by default. 
To alter this behaviour, mark  [[com.twitter.finagle.IndividualRequestTimeoutException]] as 
[[com.twitter.finagle.service.ResponseClass.RetryableFailure]] in custom 
[[com.twitter.finagle.service.ResponseClassifier]]
```

What do you think?

**Comment by mosesn** [comment]
Looks good, but remove the bit about response classification.  Finagle doesn't use response classification to decide whether to retry stuff, so you will still have to retry it manually (response classification is only for circuit breaking and stats).

**Comment by scf37** [comment]
So the right way to enable retries on timeout is to wrap built `Service[Req, Rep]` with another `TimeoutFilter` or alter `Stack` manually?

**Comment by mosesn** [comment]
Well, a RetryFilter, not a TimeoutFilter, but yes.  We're working on tooling to improve this, so it's in the pipeline.

**Comment by kevinoliver** [comment]
> We're working on tooling to improve this, so it's in the pipeline.

If you're interested, take a look at [MethodBuilder](https://github.com/twitter/finagle/blob/develop/finagle-core/src/main/scala/com/twitter/finagle/client/MethodBuilder.scala#L80). Its a work in progress, but if you look at the `withRetry` and `withTimeout` methods you can get a feel for where its going.

**Comment by scf37** [comment]
I see...
- withRetryBudget/withRetryBackoff applies only to internal Finagle errors (networking)
- ResponseClassifier has to do nothing with retries
- buil-in retries aka `RequeueFilter` are designed for network errors and failure propagation via `c.t.f.Failure` exception
- There is currently no support for application-level retries based on response classification

For example, I have Thrift service that can throw multiple exceptions, some of them retryable and some of them are not. So the *right* way is to put own RetryFilter over entire stack, including balancer?

I wonder how do you guys use this internally. Directly throw `FailureFlags` exception from business logic to signal non-retryable errors (e.g. validation) ?

**Comment by mosesn** [comment]
Yes, we put RetryFilters over everything.  Although you say, "over the loadbalancer" we typically don't encourage users at Twitter to inject filters directly into the stack, so it would also be over naming.

No, RetryPolicies can be fine-grained, so we make our RetryPolicies handle different kind of exceptions appropriately.  We are planning on using `FailureFlags` for exceptions finagle exports, but that's only to simplify parsing exceptions from finagle–you can make whatever kinds of exceptions you want, since you know what kind of complexity you want to handle.

**Comment by scf37** [comment]
So, If i get this right:
- ResponseClassifier only used by StatsFilter and `FailureAccrualFactory`
- the only built-in retries are RequeueFilter which retries only 'safe' network failures (and this can not be changed). RequeueFilter uses `context.Retries` to propagate some data over request... I didn't get how it is used though.
-  `withRetryBudget`/`withRetryBackoff` is only for RequeueFilter and, therefore, only for retrying safe network failures. Thus in real application will always be two retry filters - one for network-level retries and another for application-level retries, if applicable
- balancers ignore individual response errors and make their decisions only on service status set by `FailureAccrualFactory`
- possible Thrift/ThriftMux response are:
-- response payload
-- exception declared in thrift schema
-- Failure(Rejected, NonRetryable) - other flags are not propagated. Tests show that `Interrupted` is not propagated as well which makes me wonder.
-- TApplicationException for any other server-side exception

Is that correct?

**Comment by kevinoliver** [comment]
> ResponseClassifier only used by StatsFilter and FailureAccrualFactory

for now, yes.

> the only built-in retries are RequeueFilter which retries only 'safe' network failures (and this can not be changed). RequeueFilter uses context.Retries to propagate some data over request... I didn't get how it is used though.

Requeues can be removed from your client. Something like 

```
val client = Http.client
client.withStack(client.stack.remove(Retries.Role))
```

The Retries context is used to give the backend additional context as it is sent over the wire for HTTP and Mux. For example, a backend could decide to deprioritize retries if it chose to.

> withRetryBudget/withRetryBackoff is only for RequeueFilter and, therefore, only for retrying safe network failures. Thus in real application will always be two retry filters - one for network-level retries and another for application-level retries, if applicable

If you put your own RetryFilter on top of the client, you can share this budget. And yes, there will be two RetryFilters.

> balancers ignore individual response errors and make their decisions only on service status set by FailureAccrualFactory

Not only the status set by failure accrual, but you are correct that they directly account for response errors.

> possible Thrift/ThriftMux response are:
> -- response payload
> -- exception declared in thrift schema
> -- Failure(Rejected, NonRetryable) - other flags are not propagated. Tests show that Interrupted is not propagated as well which makes me wonder.
> -- TApplicationException for any other server-side exception

That looks right to me for ThriftMux. I'm not sure about other failure flags and would have to look at the code. I'm not sure if we have anywhere to encode the failure flags for Thrift responses.

**Comment by scf37** [comment]
`Failure.Interrupted` is converted by ThirftMux to `ServerApplicationError`. Thus, ThriftMux can also return:
- ServerError
- ServerApplicationError

It really complicates error handling. Basically, I want error handling to detect at least three error cases:
- errors that can be retried (e.g. some dependent service is unavailable or overloaded)
- errors that should not be retried (e.g. validation error)
- something unknown so it might be a bug (e.g. NPE)

ServerError/ServerApplicationError makes this really complicated. Why not let them implement `FailureFlags` at least?

Also it could be REALLY great to make `FailureFlags` public so Finagle can properly handle application exceptions, including sending correct failure flags over the wire.

**Comment by scf37** [comment]
ThriftMux connecting to unavailable port returns `ChannelWriteException`.

It seems finagle errors are diverse and unstructured. How do you write RetryPolicy for finagle clients? How do you work with persistent queues in distributed system? When processing a message from persistent you, you must clearly know whether to consume it, throw it away or put to dead letter queue.

**Comment by nepthar** [comment]
Finagle's _internal_ handling of retries definitely is less consistent than desireable and coming up to speed with it is akin to dealing with seven stages of grief. 

In fact, we have two different sections of documentations dealing with failed responses - one for [users of Finagle](https://twitter.github.io/finagle/guide/UnsuccessfulResponses.html) and one for [developers of Finagle](https://twitter.github.io/finagle/guide/developers/SignalingFailure.html).

By the time you see exceptions/errors make it out of Finagle it means that either the retry budget has been exceeded or Finagle was unable to determine that it was safe to retry "silently".

In short, what you (the user of Finagle) should do is encapsulate your retry logic in a separate RetryFilter via the methods mentioned earlier. Your retry logic should be concerned only with your exceptions and response types.

**Comment by mosesn** [comment]
@scf37 so . . . ready to make a PR? =)

**Comment by nepthar** [comment]
Now that our documentation for [MethodBuilder](https://twitter.github.io/finagle/guide/MethodBuilder.html), I'm going to mark this ticket as resolved. That provides some great guidance around retries.

Feel free to get in touch if you want to take a stab at making some documentation changes!
